crates/iceberg/src/spec/table_metadata_builder.rs

```diff
@@ -0,0 +1,2152 @@
+                    self.metadata.last_updated_ms
+                ),
+            ));
+        }
+
+        // Mutation happens in next line - must be infallible from here
+        self.changes.push(TableUpdate::AddSnapshot {
+            snapshot: snapshot.clone(),
+        });
+
+        self.metadata.last_updated_ms = snapshot.timestamp_ms();
+        self.metadata.last_sequence_number = snapshot.sequence_number();
+        self.metadata
+            .snapshots
+            .insert(snapshot.snapshot_id(), snapshot.into());
+
+        Ok(self)
+    }
+
+    /// Append a snapshot to the specified branch.
+    /// If branch is not specified, the snapshot is appended to the main branch.
+    /// The `branch` or `tag` must already exist. Retention settings from the `ref` are re-used.
+    ///
+    /// # Errors
+    /// - The ref is unknown.
+    /// - Any of the preconditions of `self.add_snapshot` are not met.
+    #[deprecated(since = "0.4.0", note = "please use `set_branch_snapshot` instead")]
+    pub fn append_snapshot(self, snapshot: Snapshot, ref_name: Option<&str>) -> Result<Self> {
+        let ref_name = ref_name.unwrap_or(MAIN_BRANCH);
+        let mut reference = self
+            .metadata
+            .refs
+            .get(ref_name)
+            .ok_or_else(|| {
+                Error::new(
+                    ErrorKind::DataInvalid,
+                    format!("Cannot append snapshot to unknown ref: '{ref_name}'"),
+                )
+            })?
+            .clone();
+
+        reference.snapshot_id = snapshot.snapshot_id();
+
+        self.add_snapshot(snapshot)?.set_ref(ref_name, reference)
+    }
+
+    /// Append a snapshot to the specified branch.
+    /// The `branch` must already exist. Retention settings from the `branch` are re-used.
+    ///
+    /// # Errors
+    /// - The branch is unknown.
+    /// - Any of the preconditions of `self.add_snapshot` are not met.
+    pub fn set_branch_snapshot(self, snapshot: Snapshot, branch: &str) -> Result<Self> {
+        let mut reference = self
+            .metadata
+            .refs
+            .get(branch)
+            .ok_or_else(|| {
+                Error::new(
+                    ErrorKind::DataInvalid,
+                    format!("Cannot append snapshot to unknown branch: '{branch}'"),
+                )
+            })?
+            .clone();
+
+        if !reference.is_branch() {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!("Cannot append snapshot to non-branch reference '{branch}'",),
+            ));
+        }
+
+        reference.snapshot_id = snapshot.snapshot_id();
+
+        self.add_snapshot(snapshot)?.set_ref(branch, reference)
+    }
+
+    /// Remove snapshots by its ids from the table metadata.
+    /// Does nothing if a snapshot id is not present.
+    /// Keeps as changes only the snapshots that were actually removed.
+    pub fn remove_snapshots(mut self, snapshot_ids: &[i64]) -> Self {
```

Comment:
Currently when removing snapshots we might have `parent_snapshot_id`s of other snapshots pointing to non-existing snapshots. This follows the behavior in Java.

Is this desirable? Or would it be more correct to set them to `None` as there is no parent available anymore?
@Fokko

Reply:
@c-thiel I think it is more correct to set it to `None`. I'll follow up on the Java side to get this fixed.